8331911: Reconsider locking for recently disarmed nmethods

neethu-prasad opened the pull request:
**Notes**
We are spending significant time on acquiring the per-nmethod as all the
threads are in same nmethod.
Adding double-check lock by calling is_armed before lock acquisition.

**Verification**

Shenendoah

> % /home/neethp/Development/opensource/jdk/build/linux-x86_64-server-release/images/jdk/bin/java -Xmx1g -Xms1g -XX:+UseShenandoahGC -Xlog:gc ManyThreadsStacks.java 2>&1 | grep "marking roots"
> [0.706s][info][gc] GC(0) Concurrent marking roots 11.519ms
> [0.752s][info][gc] GC(1) Concurrent marking roots 9.833ms
> [0.814s][info][gc] GC(2) Concurrent marking roots 10.000ms
> [0.855s][info][gc] GC(3) Concurrent marking roots 9.314ms
> [0.895s][info][gc] GC(4) Concurrent marking roots 8.937ms
> [1.213s][info][gc] GC(5) Concurrent marking roots 12.582ms
> [1.340s][info][gc] GC(6) Concurrent marking roots 9.574ms
> [1.465s][info][gc] GC(7) Concurrent marking roots 12.791ms

ZGC

> % /home/neethp/Development/opensource/jdk/build/linux-x86_64-server-release/images/jdk/bin/java -Xmx1g -Xms1g -XX:+UseShenandoahGC -Xlog:gc ManyThreadsStacks.java 2>&1 | grep "marking roots"
> [0.732s][info][gc] GC(0) Concurrent marking roots 10.694ms
> [0.782s][info][gc] GC(1) Concurrent marking roots 14.614ms
> [0.825s][info][gc] GC(2) Concurrent marking roots 12.700ms
> [0.863s][info][gc] GC(3) Concurrent marking roots 9.622ms
> [0.904s][info][gc] GC(4) Concurrent marking roots 12.892ms
> [1.244s][info][gc] GC(5) Concurrent marking roots 12.422ms
> [1.375s][info][gc] GC(6) Concurrent marking roots 12.756ms
> [1.503s][info][gc] GC(7) Concurrent marking roots 12.265ms
> [1.628s][info][gc] GC(8) Concurrent marking roots 12.309ms
> [1.754s][info][gc] GC(9) Concurrent marking roots 12.996ms
> [1.879s][info][gc] GC(10) Concurrent marking roots 9.416ms

**Issue**
https://bugs.openjdk.org/browse/JDK-8331911

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8331911](https://bugs.openjdk.org/browse/JDK-8331911): Reconsider locking for recently disarmed nmethods (**Enhancement** - P4)


### Reviewers
 * [Aleksey Shipilev](https://openjdk.org/census#shade) (@shipilev - **Reviewer**) ⚠️ Review applies to [c59034da](https://git.openjdk.org/jdk/pull/19285/files/c59034dafb81f49c6106fc6df260ac76daf87922)
 * [Erik Österlund](https://openjdk.org/census#eosterlund) (@fisk - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19285/head:pull/19285` \
`$ git checkout pull/19285`

Update a local copy of the PR: \
`$ git checkout pull/19285` \
`$ git pull https://git.openjdk.org/jdk.git pull/19285/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19285`

View PR using the GUI difftool: \
`$ git pr show -t 19285`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19285.diff">https://git.openjdk.org/jdk/pull/19285.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19285#issuecomment-2126858647)

neethu-prasad commented:
/covered

shipilev commented on src/hotspot/share/gc/shenandoah/shenandoahBarrierSetNMethod.cpp:
```diff
@@ -36,6 +36,13 @@
 bool ShenandoahBarrierSetNMethod::nmethod_entry_barrier(nmethod* nm) {
+
+  if (!is_armed(nm)) {
+    // Some other thread got here first and healed the oops
+    // and disarmed the nmethod.
```
Suggestion for the comment (here and later):

```
    // Some other thread got here first and healed the oops
    // and disarmed the nmethod. No need to continue.
```

...and then later, under the lock:

```
    // Some other thread managed to complete while we were
    // waiting for lock. No need to continue.
```

shipilev reviewed (commented):
I like this. Some stylistic comments.

shipilev reviewed (approved):
I think this is good. 

Yes, we could have restructured the code so that `nmethod_entry_barrier` was not called when nmethod is already disarmed. There are already some places where we check it externally, but the reproducer in the bug shows that it is easy to miss. So checking right here in the method looks appropriate.

@fisk might want to take a look as well.

shipilev commented:
@fisk @stefank -- are you good with this for ZGC?

fisk reviewed (changes requested):
It seems fine to me that the GC backends are responsible for checking if the nmethod is disarmed outside the lock. However, we have some callers that now check it redundantly. I think those callers should stop doing that now. Otherwise, this looks good to me.

neethu-prasad commented:
> It seems fine to me that the GC backends are responsible for checking if the nmethod is disarmed outside the lock. However, we have some callers that now check it redundantly. I think those callers should stop doing that now. Otherwise, this looks good to me.

Thanks for the feedback! Looking around the code, I think there are a few places where we can do more changes.

First, remove check here:
https://github.com/openjdk/jdk/blob/bc7d9e3d0bc663bbbeb068889082da4a9f0fa8de/src/hotspot/share/code/nmethod.cpp#L852-L855

This would force us to add the check in super-class implementation here:
https://github.com/openjdk/jdk/blob/bc7d9e3d0bc663bbbeb068889082da4a9f0fa8de/src/hotspot/share/gc/shared/barrierSetNMethod.cpp#L83

Second, we can remove the check here:
https://github.com/openjdk/jdk/blob/bc7d9e3d0bc663bbbeb068889082da4a9f0fa8de/src/hotspot/share/gc/shared/barrierSetNMethod.cpp#L175-L181

But it does not seem straightforward, because we currently skip cross-modification fence based on is_armed(...) check. Unfortunately, we cannot easily know if nmethod_entry_barrier acted or not, we only know if method is safe or not.  Can we / should we do these refactoring separately?

fisk commented:
> > It seems fine to me that the GC backends are responsible for checking if the nmethod is disarmed outside the lock. However, we have some callers that now check it redundantly. I think those callers should stop doing that now. Otherwise, this looks good to me.
> 
> 
> 
> Thanks for the feedback! Looking around the code, I think there are a few places where we can do more changes.
> 
> 
> 
> First, remove check here:
> 
> https://github.com/openjdk/jdk/blob/bc7d9e3d0bc663bbbeb068889082da4a9f0fa8de/src/hotspot/share/code/nmethod.cpp#L852-L855
> 
> 
> 
> This would force us to add the check in super-class implementation here:
> 
> https://github.com/openjdk/jdk/blob/bc7d9e3d0bc663bbbeb068889082da4a9f0fa8de/src/hotspot/share/gc/shared/barrierSetNMethod.cpp#L83
> 
> 
> 
> Second, we can remove the check here:
> 
> https://github.com/openjdk/jdk/blob/bc7d9e3d0bc663bbbeb068889082da4a9f0fa8de/src/hotspot/share/gc/shared/barrierSetNMethod.cpp#L175-L181
> 
> 
> 
> But it does not seem straightforward, because we currently skip cross-modification fence based on is_armed(...) check. Unfortunately, we cannot easily know if nmethod_entry_barrier acted or not, we only know if method is safe or not.  Can we / should we do these refactoring separately?

I see your point. However, this PR is refactoring the code to iron out who is responsible for checking is_armed, so I would prefer if we got that right in this PR. We say it should be the backend code doing that, so the callers shouldn't. I agree with all the changes you just listed and if you make them I would be happy.

Regarding the cross modifying fence, I strongly prefer to not try and be clever. Just run the cross modifying fence unconditionally after calling the backend code. We get there because the barrier was armed anyway.

neethu-prasad commented:
@fisk 
I've addressed the feedback. Can you take a look?
I did not remove the check [here](https://github.com/openjdk/jdk/blob/5dcb7a627e1cfb360719a25722588180e5de9d09/src/hotspot/share/gc/shared/barrierSetNMethod.cpp#L175-L177). Removing this check resulted in time out when `-XX:+DeoptimizeNMethodBarriersALot` flag set as it [executes deoptimization code path](https://github.com/openjdk/jdk/blob/5dcb7a627e1cfb360719a25722588180e5de9d09/src/hotspot/share/gc/shared/barrierSetNMethod.cpp#L200-L203)

fisk commented on src/hotspot/share/gc/shared/barrierSetNMethod.cpp:
```diff
@@ -172,6 +178,9 @@ int BarrierSetNMethod::nmethod_stub_entry_barrier(address* return_address_ptr) {
+  // Check for disarmed method here to avoid going into DeoptimizeNMethodBarriersALot code
+  // too often. nmethod_entry_barrier checks for disarmed status itself,
+  // but we have no visibility into whether the barrier acted or not.
   if (!bs_nm->is_armed(nm)) {
```
I would still like this check gone, together with the comment above.

shipilev replied:
I agree with Neethu, though: I think we should proceed to the `DeoptimizeNMethodBarrierALot` block and cross-modify fences only when barrier acted. We know from testing that it hurts otherwise. I would prefer this change not to introduce new performance potholes, even for fastdebug modes.

If the argument is cleanliness on who is checking "armed", and that we decide it should be solely in backend, then the middle ground might be adding the out-parameter, like `nmethod_entry_barrier(nmethod* nm, bool has_acted)`, and checking that before proceeding here? That feels uglier than just leaving the check here.

fisk commented:
> @fisk 
> 
> I've addressed the feedback. Can you take a look?
> 
> I did not remove the check [here](https://github.com/openjdk/jdk/blob/5dcb7a627e1cfb360719a25722588180e5de9d09/src/hotspot/share/gc/shared/barrierSetNMethod.cpp#L175-L177). Removing this check resulted in time out when `-XX:+DeoptimizeNMethodBarriersALot` flag set as it [executes deoptimization code path](https://github.com/openjdk/jdk/blob/5dcb7a627e1cfb360719a25722588180e5de9d09/src/hotspot/share/gc/shared/barrierSetNMethod.cpp#L200-L203)

I am quite nervous about having that silly optimization there. So I'm going to have to insist on removing it. Perhaps though, it should be fixed as a separate issue. Allow me to explain myself why it makes me nervous.
The nmethod entry barriers guard modifications to instructions and data through a mixed bag of synchronous and asynchronous cross modifying code.
Synchronous cross modifying code is the sane thing to do; we modify instructions, guarded from concurrent execution by a data flag. After modifying the instructions, the data flag is flipped, and observers are allowed to execute the instructions *after executing an instruction cross modification fence*.
On for example AArch64 we only perform synchronous cross modifying code, and the stub has the fencing machinery, so it should be okay.
However, on x86_64, we perform a mix of asynchronous and synchronous cross modifying code. The guard word is the immediate part of a compare instruction. If the new disarmed immediate is observed by concurrent execution, instruction cache coherency guarantees that we will correctly observe the cross modified instructions when they are subsequently executed.
However, when we go into the stub slow path, and check is_armed etc, these are *data reads*. That makes the dance entirely different as it suddenly performs synchronous cross modifying code. If a data read observes that the instructions have been modified, we don't have the same level of guarantees any longer, unless we perform an instruction cross modification fence.
So my concern here, is that the silly optimization to fix some verification code timeout or whatever, is in fact causing a real correctness problem for real release builds, on x86_64. By skipping the cross modification fence we perform an incomplete synchronous instruction cross modification dance that isn't sound.
Having said that, perhaps we should file a separate issue to remove that check, since it seems to fix an actual bug, while I guess this was meant more as an optimization. What do you think?

shipilev commented:
> Having said that, perhaps we should file a separate issue to remove that check, since it seems to fix an actual bug, while I guess this was meant more as an optimization. What do you think?

FTR, I don't mind executing cross-modify-fence unconditionally. I do mind going into deopts too often. I do also think that we want to stay on performance-positive side for at least an easy variant of fix, and do potentially regressing things separately. The initial motivation for this work was to resolve an issue in a service workload that runs many threads with similar stacks, and get something that we are sure net-positive and easy to reason about.

To that end, we can continue working out the final shape of the patch here, while we mitigate our current service problems with picking up a limited version of this patch with [JDK-8333716](https://bugs.openjdk.org/browse/JDK-8333716) -- it resolves only Shenandoah parts of it, though. Or, we can integrate this patch in its current form, resolving the issue on both Shenandoah and ZGC paths, and work out the check removal as the follow up of [JDK-8310239](https://bugs.openjdk.org/browse/JDK-8310239).

I think the latter alternative is more pragmatic.

fisk commented:
> > Having said that, perhaps we should file a separate issue to remove that check, since it seems to fix an actual bug, while I guess this was meant more as an optimization. What do you think?
> 
> FTR, I don't mind executing cross-modify-fence unconditionally. I do mind going into deopts too often. I do also think that we want to stay on performance-positive side for at least an easy variant of fix, and do potentially regressing things separately. The initial motivation for this work was to resolve an issue in a service workload that runs many threads with similar stacks, and get something that we are sure about for a prompt backport.

Fair enough. For what it's worth, aside for the deopt stressing option with arbitrary frequency we can update, we will not deopt more. We just perform an extra cross modifying fence when racingly entering an nmethod concurrently being disarmed. Not performing it might be slightly faster, but is a bug. But I see your point.

> To that end, we can continue working out the final shape of the patch here, while we mitigate our current service problems with picking up a limited version of this patch with [JDK-8333716](https://bugs.openjdk.org/browse/JDK-8333716) -- it resolves only Shenandoah parts of it, though. Or, we can integrate this patch in its current form, resolving the issue on both Shenandoah and ZGC paths, and work out the check removal as the follow up of [JDK-8310239](https://bugs.openjdk.org/browse/JDK-8310239).
> 
> I think the latter alternative is more pragmatic.

I'm okay with approving this patch, and we fix the actual bug separately. Sounds good? Then this is a refactoring and optimization, without the bug fix.

neethu-prasad commented:
@fisk I just merged the latest changes. Do I need approval on the merge commit or can I integrate?

fisk reviewed (approved):
Looks good. Will you file a follow-up regarding the dangerous early filtering that we discussed?

neethu-prasad commented:
Thanks for the review & approval.
I've created follow up bug - 
https://bugs.openjdk.org/browse/JDK-8334890

/integrate

shipilev commented:
/sponsor
